# 错误处理

Go语言错误处理的设计，虽然不能说完全是个失败品，但也是相当欠考虑的。Go语言并没有一个像Java一样特别完善的异常处理机制，还是使用类似C语言中，使用返回值和`if`进行判断。但是Go语言的写法还是更好看一些，因为Go支持多个返回值，所有可以把异常对象单独放到最后一个返回值里。

## error.Error()

`error`是个接口，其定义的`Error()`方法用于打印输出错误信息。下面例子调用了`strconv.Atoi()`，如果转换成功，`err`为`nil`，否则表示转换出错了，我们可以调用`Error()`方法输出其信息。

```go
func main() {
	i, err := strconv.Atoi("a")
	if err != nil{
		// 如果出错，打印错误消息，并中断处理流程
		fmt.Println("转换出错:", err.Error())
		return
	}

	fmt.Println(i)
}
```

## 抛出异常

我们定义的函数或方法内，可以通过返回值的形式指定抛出异常，调用处也可以通过这个返回值接收异常。下面是一个例子：

```go
type Person struct {
	name string
	age  int
}

func (p *Person) String() (string, error) {
	if p.name == "" && p.age == 0 {
		return "", errors.New("未初始化")
	} else {
		return fmt.Sprintf("%v / %v", p.name, p.age), nil
	}
}

func main() {
	p := &Person{}
	s, err := p.String()
	if err != nil {
		fmt.Println(err.Error())
	} else {
		fmt.Println(s)
	}
}
```

* errors.New()：该方法能将一个字符串消息包装成error类型，我们调用处的代码可以用`err.Error()`取出该消息。

## 实现`error`接口

如果需要，我们也可以实现我们自己的`error`子类用于报错。Go语言中实现接口非常简单，这个就不多介绍了。
